### ISSUE TITLE
filter bug: when cluster field is "bge-dev1", the applications with cluster "dev1" is selected

### ISSUE THREAD

**Issue by yydzhou** [argoproj/argo-cd#6337]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**
The argocd 2.0.2 has a bug on the filter function. We have two clusters, one is "dev1" and the other is "bge-dev1". However, the application with cluster "dev1" are selected when we using filter cluster "bge-dev1"

**To Reproduce**

A list of the steps required to reproduce the issue. Best of all, give us the URL to a repository that exhibits this issue.

**Expected behavior**
When filtering apps with cluster, only the applications with that cluster name should be selected. 

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by HuanxinHu** [comment]
bug is caused by the 'filterApps' function in 'applications-list.tsx' file，function use 'includes' not '==='.

```jsx
function filterApps(applications: models.Application[], pref: AppsListPreferences, search: string) {
    return applications.filter(
        app =>
            (search === '' || app.metadata.name.includes(search)) &&
            (pref.projectsFilter.length === 0 || pref.projectsFilter.includes(app.spec.project)) &&
            (pref.reposFilter.length === 0 || pref.reposFilter.includes(app.spec.source.repoURL)) &&
            (pref.syncFilter.length === 0 || pref.syncFilter.includes(app.status.sync.status)) &&
            (pref.healthFilter.length === 0 || pref.healthFilter.includes(app.status.health.status)) &&
            (pref.namespacesFilter.length === 0 || pref.namespacesFilter.some(ns => app.spec.destination.namespace && minimatch(app.spec.destination.namespace, ns))) &&
            (pref.clustersFilter.length === 0 || pref.clustersFilter.some(server => server.includes(app.spec.destination.server || app.spec.destination.name))) &&
            (pref.labelsFilter.length === 0 || pref.labelsFilter.every(selector => LabelSelector.match(selector, app.metadata.labels)))
    );
}
```

**Comment by alexmt** [comment]
Fixed. Thank you @HuanxinHu

**Comment by ciiay** [comment]
Due to the fix of #6616 this issue showed up again. After a discussion with @alexmt the below solutions for both issues were made.

- Allow fuzzy search on cluster name and exact search on cluster URL.
- Base on the starts-with of input string to create autocomplete result list. If a string starts with `https`, use exact search for cluster URL and fetch the exact cluster with URL of the input string. If a string not starts with `https`, use fuzzy search for cluster name and fetch all clusters with name which contains input string.
- Autocomplete component label and value stay the same way as now.
